### PR TITLE
JAMES-2829 delete BlobStore object storage implementation

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

**Comment by Arsnael** [comment]
```
[ERROR]   MailRepositoryContract.storingAndRemovingMessagesConcurrentlyShouldLeadToConsistentResult:474 » Execution
```
test this please

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/BlobStoreContract.java [hunk 1]
```diff
@@ -175,4 +178,51 @@ default void readShouldReturnBigSavedData() {
...
+    }
+
+    @Test
+    default void deleteShouldNotDeleteOtherBlobs() {
+        BlobId blobIdToDelete = testee().save(testee().getDefaultBucketName(), SHORT_BYTEARRAY).block();
+        BlobId otherBlobId = testee().save(testee().getDefaultBucketName(), ELEVEN_KILOBYTES).block();
+
+        testee().delete(testee().getDefaultBucketName(), blobIdToDelete).block();
+
+        InputStream read = testee().read(testee().getDefaultBucketName(), otherBlobId);
```
Comment (remk) [inline server/blob/blob-api/src/test/java/org/apache/james/blob/api/BlobStoreContract.java]: The testee().getDefaultBucketName() into a local variable could be extracted into a local variable

File: server/blob/blob-union/src/test/java/org/apache/james/blob/union/UnionBlobStoreTest.java [hunk 2]
```diff
@@ -91,6 +91,11 @@ public InputStream read(BucketName bucketName, BlobId blobId) {
             return Mono.error(new RuntimeException("broken everywhere"));
         }
 
+        @Override
+        public Mono<Void> delete(BucketName bucketName, BlobId blobId) {
+            return Mono.error(new RuntimeException("broken everywhere"));
```
Comment (remk) [inline server/blob/blob-union/src/test/java/org/apache/james/blob/union/UnionBlobStoreTest.java]: What does this message mean ?

File: server/blob/blob-api/src/test/java/org/apache/james/blob/api/BlobStoreContract.java (hunk 1 above)
Comment (Arsnael) [reply server/blob/blob-api/src/test/java/org/apache/james/blob/api/BlobStoreContract.java]: You reviewed the NO_REVIEW commit belonging to https://github.com/linagora/james-project/pull/2524, where I had that comment already and fixed it

File: server/blob/blob-union/src/test/java/org/apache/james/blob/union/UnionBlobStoreTest.java (hunk 2 above)
Comment (Arsnael) [reply server/blob/blob-union/src/test/java/org/apache/james/blob/union/UnionBlobStoreTest.java]: Reviewed the NO_REVIEW commit as well, from https://github.com/linagora/james-project/pull/2524... 

Well means it's broken, all other methods return the same, for test classes that are simulating failing BlobStores. I just kept the consistency here with the rest of those test classes. You can check the all `UnionBlobStoreTest` class if you want to see the picture :)

**Comment by Arsnael** [comment]
Rebased on the drastic changes made in https://github.com/linagora/james-project/pull/2524

**Comment by Arsnael** [comment]
```
DeleteBlobStoreContract.readShouldNotReadPartiallyWhenDeletingConcurrentlyBigBlob:163 » Execution
```
I cannot reproduce that on my local, probably that the CI is a bit weaker and needs sometimes a bit longer in duration for object storage. I notice as well that I should change the runtime exception I'm firing for the partial read case, as it's confusing with the limit of duration.

**Comment by Arsnael** [comment]
```
[ERROR] Failed to execute goal on project apache-mailet-icalendar: Could not resolve dependencies for project org.apache.james:apache-mailet-icalendar:jar:3.4.0-SNAPSHOT: Failed to collect dependencies at org.mnode.ical4j:ical4j:jar:3.0.3: Failed to read artifact descriptor for org.mnode.ical4j:ical4j:jar:3.0.3: Could not transfer artifact org.mnode.ical4j:ical4j:pom:3.0.3 from/to central (https://repo.maven.apache.org/maven2): Connection reset
```
test this please

**Comment by Arsnael** [comment]
Tests are green but : 
```
org.jenkinsci.plugins.workflow.cps.steps.ParallelStepException: Parallel step a0f7eb6f436872673101651b05b901f65b5582ac failed
[...]
Caused by: java.lang.IllegalStateException: no known node for /home/jenkins/build/workspace/572
```
Come on CI god... test this please

**Comment by trantienduchn** [comment]
Was :green_apple:

**Comment by trantienduchn** [comment]
I just want to test the stable of the concurrent reading while deleting test case. test this please

**Comment by Arsnael** [comment]
> I just want to test the stable of the concurrent reading while deleting test case. test this please

Thanks for reporting that. It seems that there is some issues indeed regarding that matter with the object storage. We need to investigate...

**Comment by Arsnael** [comment]
Rebased on master...

**Comment by Arsnael** [comment]
> > I just want to test the stable of the concurrent reading while deleting test case. test this please
> 
> Thanks for reporting that. It seems that there is some issues indeed regarding that matter with the object storage. We need to investigate...

I disabled that test at the end for scality/S3 object storage impl. I am not sure, maybe the in-memory backend docker image doesn't completely avoid that scenario, it fails randomly sometimes. I couldn't find any solution regarding this (if someone has an idea).

However I couldn't make fail the test on Swift impl, so... maybe it's just Scality?

**Comment by Arsnael** [comment]
```
[3f5d88deb305d86ccc0c53d11e23c80d7d514eb2] [ERROR] Errors: 
[3f5d88deb305d86ccc0c53d11e23c80d7d514eb2] [ERROR]   CassandraTableManagerTest » NoClassDefFound Could not initialize class org.apa...
[3f5d88deb305d86ccc0c53d11e23c80d7d514eb2] [ERROR]   CassandraTypeProviderTest » NoClassDefFound Could not initialize class org.apa...
[3f5d88deb305d86ccc0c53d11e23c80d7d514eb2] [ERROR]   CassandraTypesCreatorTest » NoClassDefFound Could not initialize class org.apa...
[3f5d88deb305d86ccc0c53d11e23c80d7d514eb2] [ERROR]   SessionWithInitializedTablesFactoryTest » NoClassDefFound Could not initialize...
[3f5d88deb305d86ccc0c53d11e23c80d7d514eb2] [ERROR]   CassandraHealthCheckTest » NoClassDefFound Could not initialize class org.apac...
[3f5d88deb305d86ccc0c53d11e23c80d7d514eb2] [ERROR]   PaggingTest » NoClassDefFound Could not initialize class org.apache.james.back...
[3f5d88deb305d86ccc0c53d11e23c80d7d514eb2] [ERROR]   CassandraSchemaVersionDAOTest » ExceptionInInitializer
```
weird... test this please

**Comment by Arsnael** [comment]
```
java.lang.NoClassDefFoundError: Could not initialize class org.apache.james.backend.rabbitmq.DockerRabbitMQSingleton
```
... test this please

File: server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAOAWSCryptoTest.java [hunk 3]
```diff
@@ -87,8 +88,8 @@ public BlobStore testee() {
     }
 
     @Override
-    @Disabled("JAMES-2829 Not supported yet")
-    public void deleteShouldPublishDeleteTimerMetrics() {
+    @Disabled("JAMES-2829 Unstable with scality/S3 impl")
```
Comment (chibenwa) [inline server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAOAWSCryptoTest.java]: Why?

Comment (trantienduchn) [reply server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAOAWSCryptoTest.java]: it's partial reading, if read & delete happen concurrent, sometime read operation just can read a part of the blob.

Comment (trantienduchn) [reply server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAOAWSCryptoTest.java]: we will find a fix and leave it disabled to have a fast merge, there are some task need this basement

Comment (Arsnael) [reply server/blob/blob-objectstorage/src/test/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAOAWSCryptoTest.java]: Seems sometimes we can read partially a blob that is being deleted concurrently. Couldn't find why yet but will try to deal with it in an other PR

**Pull request merged**